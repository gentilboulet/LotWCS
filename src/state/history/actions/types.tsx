import { ActionType } from "typesafe-actions";

import * as history from "./history";

export type IHistoryAction = ActionType<typeof history>;
