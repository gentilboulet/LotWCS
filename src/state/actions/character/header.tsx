import { createAction } from "typesafe-actions";

import * as dataArchetypes from "../../../data/archetypes";
import * as dataRanks from "../../../data/ranks";
import { IPerk } from "../../../perks";

export const setName = createAction("header/SET_NAME", (name: string) => {
  return { name };
})();

export const setConcept = createAction(
  "header/SET_CONCEPT",
  (concept: string) => {
    return { concept };
  },
)();

export const setArchetype = createAction(
  "header/SET_ARCHETYPE",
  (archetype: dataArchetypes.TArchetype) => {
    dataArchetypes.validateArchetype(archetype);
    return { archetype };
  },
)();

export const setRank = createAction(
  "header/SET_RANK",
  (rank: dataRanks.TRank) => {
    dataRanks.validateRank(rank);
    const dr = dataRanks.getRank(rank) as dataRanks.IDataRank;
    return { rank, perks: dr.perks as IPerk[] };
  },
)();
