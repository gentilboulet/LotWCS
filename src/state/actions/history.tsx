import * as constants from 'state/constants/history';

export interface IhistoryDeleteUpTo {
  type: constants.HISTORY_DELETE;
  id: number;
}

export type IHistoryAction = IhistoryDeleteUpTo;

export function historyDeleteUpTo(id: number): IhistoryDeleteUpTo {
  return { type: constants.HISTORY_DELETE, id };
}
