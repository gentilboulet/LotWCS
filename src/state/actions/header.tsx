import * as constants from 'state/constants/header';

import * as dataArchetypes from 'data/archetypes';
import * as dataRanks from 'data/ranks';

export interface IHeaderSetName {
  type: constants.HEADER_SET_NAME;
  name: string;
}

export interface IHeaderSetConcept {
  type: constants.HEADER_SET_CONCEPT;
  concept: string;
}

export interface IHeaderSetArchetype {
  type: constants.HEADER_SET_ARCHETYPE;
  archetype: string;
}

export interface IHeaderSetRank {
  type: constants.HEADER_SET_RANK;
  rank: string;
}

export type IHeaderAction = IHeaderSetName | IHeaderSetConcept | IHeaderSetArchetype | IHeaderSetRank;

export function setName(name: string): IHeaderSetName {
  return { type: constants.HEADER_SET_NAME, name };
}

export function setConcept(concept: string): IHeaderSetConcept {
  return { type: constants.HEADER_SET_CONCEPT, concept };
}

export function setArchetype(archetype: string): IHeaderSetArchetype {
  dataArchetypes.validateArchetype(archetype);
  return { archetype, type: constants.HEADER_SET_ARCHETYPE };
}

export function setRank(rank: string): IHeaderSetRank {
  dataRanks.validateRank(rank);
  return { rank, type: constants.HEADER_SET_RANK };
}
