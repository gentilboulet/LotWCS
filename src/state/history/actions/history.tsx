import { createAction } from "typesafe-actions";
import { ICharacterAction } from "../../character";

export const historyPush = createAction(
  "history/PUSH_TO_HISTORY",
  (action: ICharacterAction) => {
    return { action };
  },
)();
