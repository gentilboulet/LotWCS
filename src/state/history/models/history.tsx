import { ActionType } from "typesafe-actions";
import { ICharacterAction } from "../../character";
import * as actions from "../actions/history";

export type TInHistoryActions =
  | ICharacterAction
  | ActionType<typeof actions.resetToInitialState>;

export interface IHistoryState {
  actions: TInHistoryActions[];
}

export function initialStateFactory(): IHistoryState {
  return { actions: [actions.resetToInitialState()] };
}

export function emptyStateFactory(): IHistoryState {
  return { actions: [] };
}
