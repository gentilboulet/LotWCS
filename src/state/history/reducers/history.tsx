import { produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import { IHistoryState } from "../index";

import * as actions from "../actions/history";

export function historyReducer(
  state: IHistoryState,
  action: ActionType<typeof actions>,
): IHistoryState {
  return produce(state, draft => {
    switch (action.type) {
      case getType(actions.historyPush):
        draft.actions.push(action.payload.action);
        break;
      default:
        return;
    }
  });
}
