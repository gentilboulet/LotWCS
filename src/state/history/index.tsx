import { isActionOf } from "typesafe-actions";

import * as actions from "./actions/history";
import { IHistoryAction as a } from "./actions/types";
import { IHistoryState as s } from "./models/history";
import { historyReducer } from "./reducers/history";
export { initialStateFactory } from "./models/history";

export type IHistoryState = s;
export type IHistoryAction = a;

export function isHistoryAction(action: any): boolean {
  if (isActionOf(Object.values(actions), action)) {
    return true;
  }
  return false;
}

export function globalReducer(
  state: IHistoryState,
  action: IHistoryAction
): IHistoryState {
  return historyReducer(state, action);
}
