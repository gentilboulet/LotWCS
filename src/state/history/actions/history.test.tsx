import * as actions from "./history";

import { setName } from "../../character/actions/header";

describe("Testing history action creators", () => {
  it("should create an history push", () => {
    const toPush = setName("bob");
    const a = actions.historyPush(toPush);
    expect(a.payload.action).toBe(toPush);
    expect(a).toMatchSnapshot();
  });
});
