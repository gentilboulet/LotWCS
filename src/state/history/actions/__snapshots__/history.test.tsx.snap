// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing history action creators Testing initial state action creator should create a reset to initial state action 1`] = `
Object {
  "error": undefined,
  "meta": undefined,
  "payload": undefined,
  "type": "history/INITIAL_STATE",
}
`;

exports[`Testing history action creators should create an history delete action 1`] = `
Object {
  "error": undefined,
  "meta": undefined,
  "payload": Object {
    "id": 12,
  },
  "type": "history/DELETE_UP_TO",
}
`;
